Problem with Span<T> output

### Version

last

### What happened?

`"hello".AsSpan()`

![image](https://github.com/waf/CSharpRepl/assets/11704036/d7ff0265-f951-4e55-997a-9798f3c544a6)

Is it possible to intercept the abstract syntax tree and append a call to `.ToString()`/`.ToArray()`? That's the only way I can see this being fixed since `Span<T>` is by-design a stack-only type, and bypassing this would probably lead to undefined behavior.

The only problem with that is that not all ref-structs are so easily convertible. Take the following type for example.

```cs
ref struct A;
```

You cannot call `.ToString()` on it because that comes from `ValueType`, `A` cannot be passed into generic methods either. You would need to scaffold an entire function inside of the script context, and inject that function, which I think is a bit overkill.

Looking at other REPLs, they don't handle this at all.

Mono REPL:
```cs
csharp> Span<int>.Empty 
(1,12): error CS0029: Cannot implicitly convert type `System.Span<int>' to `object'
```

--- a/CSharpRepl.Services/CSharpRepl.Services.csproj
+++ b/CSharpRepl.Services/CSharpRepl.Services.csproj
@@ -54,6 +54,11 @@
     </AssemblyAttribute>
   </ItemGroup>
 
+
+	<ItemGroup>
+	  <EmbeddedResource Include="RuntimeHelper.cs" />
+	</ItemGroup>
+
   <ItemGroup>
     <None Update="runtime.json">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>

--- a/CSharpRepl.Services/Disassembly/Disassembler.cs
+++ b/CSharpRepl.Services/Disassembly/Disassembler.cs
@@ -27,10 +27,12 @@ internal class Disassembler
 {
     private readonly AssemblyReferenceService referenceService;
     private readonly (string name, CompileDelegate compile)[] compilers;
+    private readonly CSharpParseOptions parseOptions;
     private readonly CSharpCompilationOptions compilationOptions;
 
-    public Disassembler(CSharpCompilationOptions compilationOptions, AssemblyReferenceService referenceService, ScriptRunner scriptRunner)
+    public Disassembler(CSharpParseOptions parseOptions, CSharpCompilationOptions compilationOptions, AssemblyReferenceService referenceService, ScriptRunner scriptRunner)
     {
+        this.parseOptions = parseOptions.WithKind(SourceCodeKind.Regular);
         this.compilationOptions = compilationOptions;
         this.referenceService = referenceService;
 
@@ -45,7 +47,7 @@ internal class Disassembler
              compile: (code, optimizationLevel) => Compile(code, optimizationLevel, OutputKind.DynamicallyLinkedLibrary)),
             // Compiling as a script will work for most other cases, but it's quite verbose so we use it as a last resort.
             (name: "Scripting session (will be overly verbose)",
-             compile: (code, optimizationLevel) => scriptRunner.CompileTransient(code, optimizationLevel))
+             compile: scriptRunner.CompileTransient)
         ];
     }
 
@@ -151,7 +153,7 @@ internal class Disassembler
 
     private Compilation Compile(string code, OptimizationLevel optimizationLevel, OutputKind outputKind)
     {
-        var ast = CSharpSyntaxTree.ParseText(code, new CSharpParseOptions(LanguageVersion.Latest));
+        var ast = CSharpSyntaxTree.ParseText(code, parseOptions);
         var compilation = CSharpCompilation.Create("CompilationForDisassembly",
             [ast],
             referenceService.LoadedReferenceAssemblies,

--- a/CSharpRepl.Services/Roslyn/Formatting/CustomObjectFormatters/IEnumerableFormatter.cs
+++ b/CSharpRepl.Services/Roslyn/Formatting/CustomObjectFormatters/IEnumerableFormatter.cs
@@ -143,30 +143,27 @@ internal sealed class IEnumerableFormatter : CustomObjectFormatter<IEnumerable>
         if (type.FullName?.EndsWith(typeof(__CSharpRepl_RuntimeHelper.CharSpanOutput).FullName!) == true)
         {
             if (type.GetField(nameof(__CSharpRepl_RuntimeHelper.CharSpanOutput.Text))?.GetValue(value) is string text &&
-                type.GetField(nameof(__CSharpRepl_RuntimeHelper.CharSpanOutput.SpanWasReadOnly))?.GetValue(value) is bool readOnly)
+                type.GetField(nameof(__CSharpRepl_RuntimeHelper.CharSpanOutput.OriginalType))?.GetValue(value) is Type originalType)
             {
-                type = readOnly ? typeof(ReadOnlySpan<char>) : typeof(Span<char>);
-
-                AppendTypeName(isArray: false);
+                AppendTypeName(originalType, isArray: false);
                 sb.Append(Environment.NewLine);
                 sb.Append(formatter.FormatObjectToText(text, Level.FirstDetailed));
-                return false; 
+                return false;
             }
         }
         else if (type.FullName?.EndsWith(typeof(__CSharpRepl_RuntimeHelper.SpanOutput).FullName!) == true)
         {
-            if (type.GetField(nameof(__CSharpRepl_RuntimeHelper.SpanOutput.Array))?.GetValue(value)  is Array array &&
-                type.GetField(nameof(__CSharpRepl_RuntimeHelper.SpanOutput.SpanWasReadOnly))?.GetValue(value) is bool readOnly)
+            if (type.GetField(nameof(__CSharpRepl_RuntimeHelper.SpanOutput.OriginalType))?.GetValue(value) is Type originalType)
             {
-                type = (readOnly ? typeof(ReadOnlySpan<>) : typeof(Span<>)).MakeGenericType(array.GetType().GetElementType() ?? array.GetType());
+                type = originalType;
             }
         }
 
         var isArray = value is Array;
-        AppendTypeName(isArray);
+        AppendTypeName(type, isArray);
         return true;
 
-        void AppendTypeName(bool isArray)
+        void AppendTypeName(Type type, bool isArray)
         {
             sb.Append(
                 formatter.FormatTypeName(

--- a/CSharpRepl.Services/Roslyn/Formatting/CustomObjectFormatters/IEnumerableFormatter.cs
+++ b/CSharpRepl.Services/Roslyn/Formatting/CustomObjectFormatters/IEnumerableFormatter.cs
@@ -22,45 +22,46 @@ internal sealed class IEnumerableFormatter : CustomObjectFormatter<IEnumerable>
         var sb = new StyledStringBuilder();
 
         //header
-        AppendHeader(sb, value, formatter);
-
-        //items
-        sb.Append(" { ");
-        var enumerator = value.GetEnumerator();
-        try
+        if (AppendHeader(sb, value, formatter))
         {
-            var maxParagraphLength = LengthLimiting.GetMaxParagraphLength(level, formatter.ConsoleProfile);
-            bool any = false;
-            while (enumerator.MoveNext())
+            //items
+            sb.Append(" { ");
+            var enumerator = value.GetEnumerator();
+            try
             {
-                if (any)
+                var maxParagraphLength = LengthLimiting.GetMaxParagraphLength(level, formatter.ConsoleProfile);
+                bool any = false;
+                while (enumerator.MoveNext())
                 {
-                    sb.Append(", ");
-
-                    if (maxParagraphLength > sb.Length &&
-                        sb.Length > formatter.ConsoleProfile.Width / 2) //just heuristic
+                    if (any)
                     {
-                        sb.Append(", ...");
-                        break;
+                        sb.Append(", ");
+
+                        if (maxParagraphLength > sb.Length &&
+                            sb.Length > formatter.ConsoleProfile.Width / 2) //just heuristic
+                        {
+                            sb.Append(", ...");
+                            break;
+                        }
                     }
+                    var formattedItem = formatter.FormatObjectToText(enumerator.Current, level.Increment());
+                    sb.Append(formattedItem);
+                    any = true;
                 }
-                var formattedItem = formatter.FormatObjectToText(enumerator.Current, level.Increment());
-                sb.Append(formattedItem);
-                any = true;
             }
-        }
-        catch (Exception ex)
-        {
-            sb.Append(formatter.GetValueRetrievalExceptionText(ex, level.Increment()));
-        }
-        finally
-        {
-            if (enumerator is IDisposable disposable)
+            catch (Exception ex)
             {
-                disposable.Dispose();
+                sb.Append(formatter.GetValueRetrievalExceptionText(ex, level.Increment()));
             }
+            finally
+            {
+                if (enumerator is IDisposable disposable)
+                {
+                    disposable.Dispose();
+                }
+            }
+            sb.Append(" }");
         }
-        sb.Append(" }");
 
         return sb.ToStyledString();
     }
@@ -73,79 +74,113 @@ internal sealed class IEnumerableFormatter : CustomObjectFormatter<IEnumerable>
         }
 
         var sb = new StyledStringBuilder();
-
-        AppendHeader(sb, value, formatter);
-        var header = sb.ToStyledString().ToParagraph();
-
-        var table = new Table().AddColumns("Name", "Value", "Type");
-
-        var enumerator = value.GetEnumerator();
-        try
+        if (AppendHeader(sb, value, formatter))
         {
-            var maxItems = LengthLimiting.GetTableMaxItems(level, formatter.ConsoleProfile);
-            int counter = 0;
-            while (enumerator.MoveNext())
+            var header = sb.ToStyledString().ToParagraph();
+            var table = new Table().AddColumns("Name", "Value", "Type");
+
+            var enumerator = value.GetEnumerator();
+            try
             {
-                if (counter > maxItems)
+                var maxItems = LengthLimiting.GetTableMaxItems(level, formatter.ConsoleProfile);
+                int counter = 0;
+                while (enumerator.MoveNext())
                 {
-                    table.AddRow("...", "...", "...");
-                    break;
+                    if (counter > maxItems)
+                    {
+                        table.AddRow("...", "...", "...");
+                        break;
+                    }
+
+                    sb.Clear();
+                    sb.Append('[').Append(formatter.FormatObjectToText(counter, Level.FirstSimple)).Append(']');
+
+                    var name = sb.ToStyledString();
+
+                    var itemValue = formatter.FormatObjectToRenderable(enumerator.Current, level.Increment());
+
+                    var itemType =
+                        enumerator.Current is null ?
+                        new Paragraph("") :
+                        formatter.FormatObjectToText(enumerator.Current.GetType(), level.Increment()).ToParagraph();
+
+                    table.AddRow(name.ToParagraph(), itemValue, itemType);
+
+                    counter++;
                 }
 
-                sb.Clear();
-                sb.Append('[').Append(formatter.FormatObjectToText(counter, Level.FirstSimple)).Append(']');
-
-                var name = sb.ToStyledString();
-
-                var itemValue = formatter.FormatObjectToRenderable(enumerator.Current, level.Increment());
-
-                var itemType =
-                    enumerator.Current is null ?
-                    new Paragraph("") :
-                    formatter.FormatObjectToText(enumerator.Current.GetType(), level.Increment()).ToParagraph();
-
-                table.AddRow(name.ToParagraph(), itemValue, itemType);
-
-                counter++;
+                if (counter == 0)
+                {
+                    return new FormattedObjectRenderable(header, renderOnNewLine: false);
+                }
             }
-
-            if (counter == 0)
+            catch (Exception ex)
             {
-                return new FormattedObjectRenderable(header, renderOnNewLine: false);
+                table.AddRow(new Paragraph(""), formatter.GetValueRetrievalExceptionText(ex, level.Increment()).ToParagraph(), new Paragraph(""));
             }
-        }
-        catch (Exception ex)
-        {
-            table.AddRow(new Paragraph(""), formatter.GetValueRetrievalExceptionText(ex, level.Increment()).ToParagraph(), new Paragraph(""));
-        }
-        finally
-        {
-            if (enumerator is IDisposable disposable)
+            finally
             {
-                disposable.Dispose();
+                if (enumerator is IDisposable disposable)
+                {
+                    disposable.Dispose();
+                }
             }
-        }
 
-        return new FormattedObjectRenderable(
-            new RenderableSequence(header, table, separateByLineBreak: true),
-            renderOnNewLine: false);
+            return new FormattedObjectRenderable(
+                new RenderableSequence(header, table, separateByLineBreak: true),
+                renderOnNewLine: false);
+        }
+        else
+        {
+            return new FormattedObjectRenderable(sb.ToStyledString().ToParagraph(), renderOnNewLine: false);
+        }
     }
 
-    private static void AppendHeader(StyledStringBuilder sb, IEnumerable value, Formatter formatter)
+    private static bool AppendHeader(StyledStringBuilder sb, IEnumerable value, Formatter formatter)
     {
-        var isArray = value is Array;
-        sb.Append(
-            formatter.FormatTypeName(
-                isArray ? (value.GetType().GetElementType() ?? value.GetType()) : value.GetType(),
-                showNamespaces: false,
-                useLanguageKeywords: true,
-                hideSystemNamespace: true));
+        var type = value.GetType();
 
-        if (TryGetCount(value, formatter, out var count))
+        if (type.FullName?.EndsWith(typeof(__CSharpRepl_RuntimeHelper.CharSpanOutput).FullName!) == true)
         {
-            sb.Append(isArray ? '[' : '(');
-            sb.Append(count);
-            sb.Append(isArray ? ']' : ')');
+            if (type.GetField(nameof(__CSharpRepl_RuntimeHelper.CharSpanOutput.Text))?.GetValue(value) is string text &&
+                type.GetField(nameof(__CSharpRepl_RuntimeHelper.CharSpanOutput.SpanWasReadOnly))?.GetValue(value) is bool readOnly)
+            {
+                type = readOnly ? typeof(ReadOnlySpan<char>) : typeof(Span<char>);
+
+                AppendTypeName(isArray: false);
+                sb.Append(Environment.NewLine);
+                sb.Append(formatter.FormatObjectToText(text, Level.FirstDetailed));
+                return false; 
+            }
+        }
+        else if (type.FullName?.EndsWith(typeof(__CSharpRepl_RuntimeHelper.SpanOutput).FullName!) == true)
+        {
+            if (type.GetField(nameof(__CSharpRepl_RuntimeHelper.SpanOutput.Array))?.GetValue(value)  is Array array &&
+                type.GetField(nameof(__CSharpRepl_RuntimeHelper.SpanOutput.SpanWasReadOnly))?.GetValue(value) is bool readOnly)
+            {
+                type = (readOnly ? typeof(ReadOnlySpan<>) : typeof(Span<>)).MakeGenericType(array.GetType().GetElementType() ?? array.GetType());
+            }
+        }
+
+        var isArray = value is Array;
+        AppendTypeName(isArray);
+        return true;
+
+        void AppendTypeName(bool isArray)
+        {
+            sb.Append(
+                formatter.FormatTypeName(
+                    isArray ? (type.GetElementType() ?? type) : type,
+                    showNamespaces: false,
+                    useLanguageKeywords: true,
+                    hideSystemNamespace: true));
+
+            if (TryGetCount(value, formatter, out var count))
+            {
+                sb.Append(isArray ? '[' : '(');
+                sb.Append(count);
+                sb.Append(isArray ? ']' : ')');
+            }
         }
     }
 

--- a/CSharpRepl.Services/Roslyn/Formatting/FormattedObject.cs
+++ b/CSharpRepl.Services/Roslyn/Formatting/FormattedObject.cs
@@ -1,5 +1,4 @@
-﻿using System;
-using System.Collections.Generic;
+﻿using System.Collections.Generic;
 using Spectre.Console.Rendering;
 
 namespace CSharpRepl.Services.Roslyn.Formatting;

--- a/CSharpRepl.Services/Roslyn/References/AssemblyReferenceService.cs
+++ b/CSharpRepl.Services/Roslyn/References/AssemblyReferenceService.cs
@@ -34,14 +34,16 @@ internal sealed class AssemblyReferenceService
     private readonly HashSet<string> implementationAssemblyPaths;
     private readonly HashSet<string> sharedFrameworkImplementationAssemblyPaths;
     private readonly HashSet<UsingDirectiveSyntax> usings;
+    private readonly CSharpParseOptions parseOptions;
 
     public IReadOnlySet<string> ImplementationAssemblyPaths => implementationAssemblyPaths;
     public IReadOnlySet<MetadataReference> LoadedImplementationAssemblies => loadedImplementationAssemblies;
     public IReadOnlySet<MetadataReference> LoadedReferenceAssemblies => loadedReferenceAssemblies;
     public IReadOnlyCollection<UsingDirectiveSyntax> Usings => usings;
 
-    public AssemblyReferenceService(Configuration config, ITraceLogger logger)
+    public AssemblyReferenceService(Configuration config, CSharpParseOptions parseOptions, ITraceLogger logger)
     {
+        this.parseOptions = parseOptions;
         this.dotnetInstallationLocator = new DotNetInstallationLocator(logger);
         this.referenceAssemblyPaths = [];
         this.implementationAssemblyPaths = [];
@@ -201,7 +203,7 @@ internal sealed class AssemblyReferenceService
     }
 
     internal IReadOnlyCollection<UsingDirectiveSyntax> GetUsings(string code) =>
-        CSharpSyntaxTree.ParseText(code)
+        CSharpSyntaxTree.ParseText(code, parseOptions)
             .GetRoot()
             .DescendantNodes()
             .OfType<UsingDirectiveSyntax>()
@@ -210,7 +212,7 @@ internal sealed class AssemblyReferenceService
     internal void TrackUsings(IReadOnlyCollection<UsingDirectiveSyntax> usingsToAdd) =>
         usings.UnionWith(usingsToAdd);
 
-    private IReadOnlyCollection<MetadataReference> CreateDefaultReferences(string assemblyPath, IReadOnlyCollection<string> assemblies)
+    private List<PortableExecutableReference> CreateDefaultReferences(string assemblyPath, IReadOnlyCollection<string> assemblies)
     {
         return assemblies
             .AsParallel()

--- a/CSharpRepl.Services/Roslyn/RoslynServices.cs
+++ b/CSharpRepl.Services/Roslyn/RoslynServices.cs
@@ -465,24 +465,54 @@ public sealed partial class RoslynServices
         if ((await scriptRunner.HasValueReturningStatement(input, cancellationToken).ConfigureAwait(false)).TryGet(out var result) &&
             result.Type.IsRefLikeType)
         {
+            var root = result.Expression.SyntaxTree.GetRoot(cancellationToken);
+            var expressionToBeWrapped = result.Expression;
+            ExpressionSyntax wrappedExpression;
             if (result.Type is { Name: "Span" or "ReadOnlySpan", ContainingNamespace.Name: "System" })
             {
-                var root = result.Expression.SyntaxTree.GetRoot(cancellationToken);
-                var wrappedExpression =
-                SyntaxFactory.InvocationExpression(
-                        SyntaxFactory.MemberAccessExpression(
-                            SyntaxKind.SimpleMemberAccessExpression,
-                            SyntaxFactory.IdentifierName(nameof(__CSharpRepl_RuntimeHelper)),
-                            SyntaxFactory.IdentifierName(nameof(__CSharpRepl_RuntimeHelper.HandleSpanOutput))))
-                    .WithArgumentList(
-                        SyntaxFactory.ArgumentList(
-                            SyntaxFactory.SingletonSeparatedList(
-                                SyntaxFactory.Argument(
-                                    result.Expression))));
-                root = root.ReplaceNode(result.Expression, wrappedExpression);
-                return root.GetText().ToString();
+                wrappedExpression = Wrap(nameof(__CSharpRepl_RuntimeHelper.HandleSpanOutput), result.Expression);
             }
+            else
+            {
+                var toStringMethod = result.Type
+                    .GetMembers(nameof(ToString))
+                    .OfType<IMethodSymbol>()
+                    .Where(m => m.ReturnType.SpecialType == SpecialType.System_String && m.Parameters.Length == 0 && m.IsOverride)
+                    .FirstOrDefault();
+
+                if (toStringMethod != null)
+                {
+                    expressionToBeWrapped = SyntaxFactory.InvocationExpression(
+                        SyntaxFactory.MemberAccessExpression(
+                            SyntaxKind.SimpleMemberAccessExpression, 
+                            expressionToBeWrapped, 
+                            SyntaxFactory.IdentifierName(nameof(ToString))));
+                }
+                else
+                {
+                    expressionToBeWrapped = SyntaxFactory.LiteralExpression(
+                        SyntaxKind.StringLiteralExpression, 
+                        SyntaxFactory.Literal($"Cannot output a value of '{result.Type.Name}' because it's a ref-struct. It has to override ToString() to see its value."));
+                }
+                wrappedExpression = Wrap(
+                    nameof(__CSharpRepl_RuntimeHelper.HandleRefStructOutput),
+                    expressionToBeWrapped);
+            }
+            root = root.ReplaceNode(result.Expression, wrappedExpression);
+            return root.GetText().ToString();
         }
         return input;
+
+        static ExpressionSyntax Wrap(string runtimeHelperMethod, ExpressionSyntax expressionToBeWrapped) =>
+             SyntaxFactory.InvocationExpression(
+                            SyntaxFactory.MemberAccessExpression(
+                                SyntaxKind.SimpleMemberAccessExpression,
+                                SyntaxFactory.IdentifierName(nameof(__CSharpRepl_RuntimeHelper)),
+                                SyntaxFactory.IdentifierName(runtimeHelperMethod)))
+                        .WithArgumentList(
+                            SyntaxFactory.ArgumentList(
+                                SyntaxFactory.SingletonSeparatedList(
+                                    SyntaxFactory.Argument(
+                                        expressionToBeWrapped))));
     }
 }

--- a/CSharpRepl.Services/Roslyn/RoslynServices.cs
+++ b/CSharpRepl.Services/Roslyn/RoslynServices.cs
@@ -47,6 +47,7 @@ public sealed partial class RoslynServices
     private readonly SemaphoreSlim semaphore = new(1);
     private readonly IPromptCallbacks defaultPromptCallbacks = new PromptCallbacks();
     private readonly ThreadLocal<OverloadItemGenerator> overloadItemGenerator;
+    private readonly CSharpParseOptions parseOptions = CSharpParseOptions.Default.WithKind(SourceCodeKind.Script).WithLanguageVersion(LanguageVersion.Latest);
     private ScriptRunner? scriptRunner;
     private WorkspaceManager? workspaceManager;
     private Disassembler? disassembler;
@@ -78,7 +79,7 @@ public sealed partial class RoslynServices
         {
             logger.Log("Starting background initialization");
 
-            this.referenceService = new AssemblyReferenceService(config, logger);
+            this.referenceService = new AssemblyReferenceService(config, parseOptions, logger);
 
             this.compilationOptions = new CSharpCompilationOptions(
                 OutputKind.DynamicallyLinkedLibrary,
@@ -91,9 +92,9 @@ public sealed partial class RoslynServices
             // is updated alongside. The workspace is a datamodel used in "editor services" like
             // syntax highlighting, autocompletion, and roslyn symbol queries.
             this.workspaceManager = new WorkspaceManager(compilationOptions, referenceService, logger);
-            this.scriptRunner = new ScriptRunner(workspaceManager, compilationOptions, referenceService, console, config);
+            this.scriptRunner = new ScriptRunner(workspaceManager, parseOptions, compilationOptions, referenceService, console, config);
 
-            this.disassembler = new Disassembler(compilationOptions, referenceService, scriptRunner);
+            this.disassembler = new Disassembler(parseOptions, compilationOptions, referenceService, scriptRunner);
             this.prettyPrinter = new PrettyPrinter(console, highlighter, config);
             this.symbolExplorer = new SymbolExplorer(referenceService, scriptRunner);
             this.autocompleteService = new AutoCompleteService(highlighter, cache, config);

--- a/CSharpRepl.Services/Roslyn/RoslynServices.cs
+++ b/CSharpRepl.Services/Roslyn/RoslynServices.cs
@@ -6,7 +6,9 @@ using System;
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
+using System.IO;
 using System.Linq;
+using System.Reflection;
 using System.Threading;
 using System.Threading.Tasks;
 using CSharpRepl.Services.Completion;
@@ -113,7 +115,7 @@ public sealed partial class RoslynServices
             //each RunCompilation (modifies script state) and UpdateCurrentDocument (changes CurrentDocument) cannot be run concurrently
             await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
 
-            input = input.Trim();
+            input = await InterceptInputWhenRefStructReturnNeedsToBeHandled(input.Trim(), cancellationToken).ConfigureAwait(false);
             EvaluatingInput?.Invoke(input);
 
             var result = await scriptRunner
@@ -427,7 +429,22 @@ public sealed partial class RoslynServices
 
             logger.Log("Warm-up Starting");
 
-            var evaluationTask = EvaluateAsync(@"_ = ""REPL Warmup""", args);
+            const string RuntimeHelperName = "CSharpRepl.Services.RuntimeHelper.cs";
+            Task evaluationTask;
+            using (var runtimeHelperStream = Assembly.GetExecutingAssembly().GetManifestResourceStream(RuntimeHelperName))
+            {
+                if (runtimeHelperStream != null)
+                {
+                    var runtimeHelperText = new StreamReader(runtimeHelperStream).ReadToEnd();
+                    evaluationTask = EvaluateAsync(runtimeHelperText, args);
+                }
+                else
+                {
+                    evaluationTask = Task.CompletedTask;
+                    console.WriteErrorLine($"Unable to load '{RuntimeHelperName}'");
+                }
+            }
+
             var highlightTask = SyntaxHighlightAsync(@"_ = ""REPL Warmup""");
             var formattingTask = FormatInput(@"_=""REPL Warmup"";", caret: 16, formatParentNodeOnly: false, default);
             var completionTask = Task.WhenAny(
@@ -440,4 +457,32 @@ public sealed partial class RoslynServices
             await Task.WhenAll(evaluationTask, highlightTask, formattingTask, completionTask).ConfigureAwait(false);
             logger.Log("Warm-up Complete");
         });
+
+    private async Task<string> InterceptInputWhenRefStructReturnNeedsToBeHandled(string input, CancellationToken cancellationToken)
+    {
+        await Initialization.ConfigureAwait(false);
+
+        if ((await scriptRunner.HasValueReturningStatement(input, cancellationToken).ConfigureAwait(false)).TryGet(out var result) &&
+            result.Type.IsRefLikeType)
+        {
+            if (result.Type is { Name: "Span" or "ReadOnlySpan", ContainingNamespace.Name: "System" })
+            {
+                var root = result.Expression.SyntaxTree.GetRoot(cancellationToken);
+                var wrappedExpression =
+                SyntaxFactory.InvocationExpression(
+                        SyntaxFactory.MemberAccessExpression(
+                            SyntaxKind.SimpleMemberAccessExpression,
+                            SyntaxFactory.IdentifierName(nameof(__CSharpRepl_RuntimeHelper)),
+                            SyntaxFactory.IdentifierName(nameof(__CSharpRepl_RuntimeHelper.HandleSpanOutput))))
+                    .WithArgumentList(
+                        SyntaxFactory.ArgumentList(
+                            SyntaxFactory.SingletonSeparatedList(
+                                SyntaxFactory.Argument(
+                                    result.Expression))));
+                root = root.ReplaceNode(result.Expression, wrappedExpression);
+                return root.GetText().ToString();
+            }
+        }
+        return input;
+    }
 }

--- a/CSharpRepl.Services/Roslyn/RoslynServices.cs
+++ b/CSharpRepl.Services/Roslyn/RoslynServices.cs
@@ -462,17 +462,20 @@ public sealed partial class RoslynServices
     {
         await Initialization.ConfigureAwait(false);
 
-        if ((await scriptRunner.HasValueReturningStatement(input, cancellationToken).ConfigureAwait(false)).TryGet(out var result) &&
-            result.Type.IsRefLikeType)
+        if ((await scriptRunner.HasValueReturningStatement(input, cancellationToken).ConfigureAwait(false)).TryGet(out var result))
         {
             var root = result.Expression.SyntaxTree.GetRoot(cancellationToken);
             var expressionToBeWrapped = result.Expression;
-            ExpressionSyntax wrappedExpression;
+            var wrappedExpression = expressionToBeWrapped;
             if (result.Type is { Name: "Span" or "ReadOnlySpan", ContainingNamespace.Name: "System" })
             {
                 wrappedExpression = Wrap(nameof(__CSharpRepl_RuntimeHelper.HandleSpanOutput), result.Expression);
             }
-            else
+            else if (result.Type is { Name: "Memory" or "ReadOnlyMemory", ContainingNamespace.Name: "System" })
+            {
+                wrappedExpression = Wrap(nameof(__CSharpRepl_RuntimeHelper.HandleMemoryOutput), result.Expression);
+            }
+            else if (result.Type.IsRefLikeType)
             {
                 var toStringMethod = result.Type
                     .GetMembers(nameof(ToString))
@@ -484,22 +487,26 @@ public sealed partial class RoslynServices
                 {
                     expressionToBeWrapped = SyntaxFactory.InvocationExpression(
                         SyntaxFactory.MemberAccessExpression(
-                            SyntaxKind.SimpleMemberAccessExpression, 
-                            expressionToBeWrapped, 
+                            SyntaxKind.SimpleMemberAccessExpression,
+                            expressionToBeWrapped,
                             SyntaxFactory.IdentifierName(nameof(ToString))));
                 }
                 else
                 {
                     expressionToBeWrapped = SyntaxFactory.LiteralExpression(
-                        SyntaxKind.StringLiteralExpression, 
+                        SyntaxKind.StringLiteralExpression,
                         SyntaxFactory.Literal($"Cannot output a value of '{result.Type.Name}' because it's a ref-struct. It has to override ToString() to see its value."));
                 }
                 wrappedExpression = Wrap(
                     nameof(__CSharpRepl_RuntimeHelper.HandleRefStructOutput),
                     expressionToBeWrapped);
             }
-            root = root.ReplaceNode(result.Expression, wrappedExpression);
-            return root.GetText().ToString();
+
+            if (result.Expression != wrappedExpression)
+            {
+                root = root.ReplaceNode(result.Expression, wrappedExpression);
+                return root.GetText().ToString();
+            }
         }
         return input;
 

--- a/CSharpRepl.Services/Roslyn/Scripting/ScriptRunner.cs
+++ b/CSharpRepl.Services/Roslyn/Scripting/ScriptRunner.cs
@@ -123,7 +123,7 @@ internal sealed class ScriptRunner
 
     private async Task<EvaluationResult.Success> CreateSuccessfulResult(string text, ScriptState<object> state, CancellationToken cancellationToken)
     {
-        var hasValueReturningStatement = await HasValueReturningStatement(text, cancellationToken).ConfigureAwait(false);
+        var hasValueReturningStatement = (await HasValueReturningStatement(text, cancellationToken).ConfigureAwait(false)).HasValue;
 
         referenceAssemblyService.AddImplementationAssemblyReferences(state.Script.GetCompilation().References);
         var frameworkReferenceAssemblies = referenceAssemblyService.LoadedReferenceAssemblies;
@@ -145,7 +145,7 @@ internal sealed class ScriptRunner
         return await scriptTask.ConfigureAwait(false);
     }
 
-    private async Task<bool> HasValueReturningStatement(string text, CancellationToken cancellationToken)
+    internal async Task<(ExpressionSyntax Expression, ITypeSymbol Type)?> HasValueReturningStatement(string text, CancellationToken cancellationToken)
     {
         var sourceText = SourceText.From(text);
         var document = workspaceManager.CurrentDocument.WithText(sourceText);
@@ -160,11 +160,14 @@ internal sealed class ScriptRunner
             var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
             if (semanticModel != null)
             {
-                var typeInfo = semanticModel.GetTypeInfo(possiblyValueReturningStatement.Expression, cancellationToken);
-                return typeInfo.ConvertedType?.SpecialType != SpecialType.System_Void;
+                var returnType = semanticModel.GetTypeInfo(possiblyValueReturningStatement.Expression, cancellationToken).ConvertedType;
+                if (returnType?.SpecialType is not (null or SpecialType.System_Void))
+                {
+                    return (possiblyValueReturningStatement.Expression, returnType);
+                }
             }
         }
-        return false;
+        return null;
     }
 
     private ScriptGlobals CreateGlobalsObject(string[]? args)

--- a/CSharpRepl.Services/Roslyn/Scripting/ScriptRunner.cs
+++ b/CSharpRepl.Services/Roslyn/Scripting/ScriptRunner.cs
@@ -29,12 +29,14 @@ internal sealed class ScriptRunner
     private readonly CompositeAlternativeReferenceResolver alternativeReferenceResolver;
     private readonly MetadataReferenceResolver metadataResolver;
     private readonly WorkspaceManager workspaceManager;
+    private readonly CSharpParseOptions parseOptions;
     private readonly AssemblyReferenceService referenceAssemblyService;
     private ScriptOptions scriptOptions;
     private ScriptState<object>? state;
 
     public ScriptRunner(
         WorkspaceManager workspaceManager,
+        CSharpParseOptions parseOptions,
         CSharpCompilationOptions compilationOptions,
         AssemblyReferenceService referenceAssemblyService,
         IConsoleEx console,
@@ -42,6 +44,7 @@ internal sealed class ScriptRunner
     {
         this.console = console;
         this.workspaceManager = workspaceManager;
+        this.parseOptions = parseOptions;
         this.referenceAssemblyService = referenceAssemblyService;
         this.assemblyLoader = new InteractiveAssemblyLoader(new MetadataShadowCopyProvider());
 
@@ -110,7 +113,7 @@ internal sealed class ScriptRunner
     {
         return CSharpCompilation.CreateScriptCompilation(
             "CompilationTransient",
-            CSharpSyntaxTree.ParseText(code, CSharpParseOptions.Default.WithKind(SourceCodeKind.Script).WithLanguageVersion(LanguageVersion.Latest)),
+            CSharpSyntaxTree.ParseText(code, parseOptions),
             scriptOptions.MetadataReferences,
             new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary, usings: scriptOptions.Imports, optimizationLevel: optimizationLevel, allowUnsafe: scriptOptions.AllowUnsafe, metadataReferenceResolver: metadataResolver),
             previousScriptCompilation: state?.Script.GetCompilation() is CSharpCompilation previous ? previous : null,
@@ -165,7 +168,5 @@ internal sealed class ScriptRunner
     }
 
     private ScriptGlobals CreateGlobalsObject(string[]? args)
-    {
-        return new ScriptGlobals(console, args ?? []);
-    }
+        => new(console, args ?? []);
 }

--- a/CSharpRepl.Services/RuntimeHelper.cs
+++ b/CSharpRepl.Services/RuntimeHelper.cs
@@ -6,19 +6,25 @@
 
 public static class __CSharpRepl_RuntimeHelper
 {
-    public static SpanOutput HandleSpanOutput<T>(System.Span<T> span) => SpanOutput.Create<T>(span, false);
-    public static SpanOutput HandleSpanOutput<T>(System.ReadOnlySpan<T> span) => SpanOutput.Create(span, true);
+    public static SpanOutput HandleSpanOutput<T>(System.Span<T> span) => SpanOutput.Create<T>(span, typeof(System.Span<T>));
+    public static SpanOutput HandleSpanOutput<T>(System.ReadOnlySpan<T> span) => SpanOutput.Create(span, typeof(System.ReadOnlySpan<T>));
 
-    public static CharSpanOutput HandleSpanOutput(System.Span<char> span) => CharSpanOutput.Create(span, false);
-    public static CharSpanOutput HandleSpanOutput(System.ReadOnlySpan<char> span) => CharSpanOutput.Create(span, true);
+    public static CharSpanOutput HandleSpanOutput(System.Span<char> span) => CharSpanOutput.Create(span, typeof(System.Span<char>));
+    public static CharSpanOutput HandleSpanOutput(System.ReadOnlySpan<char> span) => CharSpanOutput.Create(span, typeof(System.ReadOnlySpan<char>));
+
+    public static SpanOutput HandleMemoryOutput<T>(System.Memory<T> memory) => SpanOutput.Create<T>(memory.Span, typeof(System.Memory<T>));
+    public static SpanOutput HandleMemoryOutput<T>(System.ReadOnlyMemory<T> memory) => SpanOutput.Create(memory.Span, typeof(System.ReadOnlyMemory<T>));
+
+    public static CharSpanOutput HandleMemoryOutput(System.Memory<char> memory) => CharSpanOutput.Create(memory.Span, typeof(System.Memory<char>));
+    public static CharSpanOutput HandleMemoryOutput(System.ReadOnlyMemory<char> memory) => CharSpanOutput.Create(memory.Span, typeof(System.ReadOnlyMemory<char>));
 
     public static RefStructOutput HandleRefStructOutput(string text) => new(text);
 
-    public abstract class SpanOutputBase(int originalLength, bool spanWasReadOnly) 
+    public abstract class SpanOutputBase(int originalLength, System.Type originalType)
         : System.Collections.IEnumerable
     {
         public readonly int OriginalLength = originalLength;
-        public readonly bool SpanWasReadOnly = spanWasReadOnly;
+        public readonly System.Type OriginalType = originalType;
 
         //Necessary for correct output formatting.
         public int Count => OriginalLength;
@@ -26,23 +32,23 @@ public static class __CSharpRepl_RuntimeHelper
         public abstract System.Collections.IEnumerator GetEnumerator();
     }
 
-    public sealed class SpanOutput(System.Array array, int originalLength, bool spanWasReadOnly) 
-        : SpanOutputBase(originalLength, spanWasReadOnly)
+    public sealed class SpanOutput(System.Array array, int originalLength, System.Type originalType)
+        : SpanOutputBase(originalLength, originalType)
     {
         private const int MaxLength = 1024;
 
-        public readonly System.Array Array = array;
+        private readonly System.Array array = array;
 
-        public override System.Collections.IEnumerator GetEnumerator() => Array.GetEnumerator();
+        public override System.Collections.IEnumerator GetEnumerator() => array.GetEnumerator();
 
-        public static SpanOutput Create<T>(System.ReadOnlySpan<T> span, bool spanWasReadOnly) => new(
+        public static SpanOutput Create<T>(System.ReadOnlySpan<T> span, System.Type originalType) => new(
             span[..System.Math.Min(MaxLength, span.Length)].ToArray(),
             span.Length,
-            spanWasReadOnly);
+            originalType);
     }
 
-    public sealed class CharSpanOutput(string text, int originalLength, bool spanWasReadOnly) 
-        : SpanOutputBase(originalLength, spanWasReadOnly)
+    public sealed class CharSpanOutput(string text, int originalLength, System.Type originalType)
+        : SpanOutputBase(originalLength, originalType)
     {
         private const int MaxLength = 10_000;
 
@@ -50,7 +56,7 @@ public static class __CSharpRepl_RuntimeHelper
 
         public override System.Collections.IEnumerator GetEnumerator() => Text.GetEnumerator();
 
-        public static CharSpanOutput Create(System.ReadOnlySpan<char> span, bool spanWasReadOnly)
+        public static CharSpanOutput Create(System.ReadOnlySpan<char> span, System.Type originalType)
         {
             var len = System.Math.Min(MaxLength, span.Length);
             System.Span<char> buffer = stackalloc char[len];
@@ -61,7 +67,7 @@ public static class __CSharpRepl_RuntimeHelper
                 buffer[^2] = '.';
                 buffer[^3] = '.';
             }
-            return new(buffer.ToString(), span.Length, spanWasReadOnly);
+            return new(buffer.ToString(), span.Length, originalType);
         }
     }
 

--- a/CSharpRepl.Services/RuntimeHelper.cs
+++ b/CSharpRepl.Services/RuntimeHelper.cs
@@ -1,0 +1,65 @@
+﻿// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//This file loaded from resources and injected into CSharpRepl session in the warm-up phase.
+
+public static class __CSharpRepl_RuntimeHelper
+{
+    public static SpanOutput HandleSpanOutput<T>(System.Span<T> span) => SpanOutput.Create<T>(span, false);
+    public static SpanOutput HandleSpanOutput<T>(System.ReadOnlySpan<T> span) => SpanOutput.Create(span, true);
+
+    public static CharSpanOutput HandleSpanOutput(System.Span<char> span) => CharSpanOutput.Create(span, false);
+    public static CharSpanOutput HandleSpanOutput(System.ReadOnlySpan<char> span) => CharSpanOutput.Create(span, true);
+
+    public abstract class SpanOutputBase(int originalLength, bool spanWasReadOnly) 
+        : System.Collections.IEnumerable
+    {
+        public readonly int OriginalLength = originalLength;
+        public readonly bool SpanWasReadOnly = spanWasReadOnly;
+
+        //Necessary for correct output formatting.
+        public int Count => OriginalLength;
+
+        public abstract System.Collections.IEnumerator GetEnumerator();
+    }
+
+    public sealed class SpanOutput(System.Array array, int originalLength, bool spanWasReadOnly) 
+        : SpanOutputBase(originalLength, spanWasReadOnly)
+    {
+        private const int MaxLength = 1024;
+
+        public readonly System.Array Array = array;
+
+        public override System.Collections.IEnumerator GetEnumerator() => Array.GetEnumerator();
+
+        public static SpanOutput Create<T>(System.ReadOnlySpan<T> span, bool spanWasReadOnly) => new(
+            span[..System.Math.Min(MaxLength, span.Length)].ToArray(),
+            span.Length,
+            spanWasReadOnly);
+    }
+
+    public sealed class CharSpanOutput(string text, int originalLength, bool spanWasReadOnly) 
+        : SpanOutputBase(originalLength, spanWasReadOnly)
+    {
+        private const int MaxLength = 10_000;
+
+        public readonly string Text = text;
+
+        public override System.Collections.IEnumerator GetEnumerator() => Text.GetEnumerator();
+
+        public static CharSpanOutput Create(System.ReadOnlySpan<char> span, bool spanWasReadOnly)
+        {
+            var len = System.Math.Min(MaxLength, span.Length);
+            System.Span<char> buffer = stackalloc char[len];
+            span[..len].CopyTo(buffer);
+            if (span.Length > len)
+            {
+                buffer[^1] = '.';
+                buffer[^2] = '.';
+                buffer[^3] = '.';
+            }
+            return new(buffer.ToString(), span.Length, spanWasReadOnly);
+        }
+    }
+}

--- a/CSharpRepl.Services/RuntimeHelper.cs
+++ b/CSharpRepl.Services/RuntimeHelper.cs
@@ -12,6 +12,8 @@ public static class __CSharpRepl_RuntimeHelper
     public static CharSpanOutput HandleSpanOutput(System.Span<char> span) => CharSpanOutput.Create(span, false);
     public static CharSpanOutput HandleSpanOutput(System.ReadOnlySpan<char> span) => CharSpanOutput.Create(span, true);
 
+    public static RefStructOutput HandleRefStructOutput(string text) => new(text);
+
     public abstract class SpanOutputBase(int originalLength, bool spanWasReadOnly) 
         : System.Collections.IEnumerable
     {
@@ -61,5 +63,11 @@ public static class __CSharpRepl_RuntimeHelper
             }
             return new(buffer.ToString(), span.Length, spanWasReadOnly);
         }
+    }
+
+    public class RefStructOutput(string text)
+    {
+        private readonly string text = text;
+        public override string ToString() => text;
     }
 }

--- a/CSharpRepl.Tests/CompletionTests.cs
+++ b/CSharpRepl.Tests/CompletionTests.cs
@@ -2,7 +2,6 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-using System;
 using System.Linq;
 using System.Threading.Tasks;
 using CSharpRepl.PrettyPromptConfig;

--- a/CSharpRepl.Tests/DisassemblerTests.cs
+++ b/CSharpRepl.Tests/DisassemblerTests.cs
@@ -1,5 +1,4 @@
-﻿using System;
-using System.IO;
+﻿using System.IO;
 using System.Threading.Tasks;
 using CSharpRepl.Services;
 using CSharpRepl.Services.Roslyn;

--- a/CSharpRepl.Tests/EvaluationTests.cs
+++ b/CSharpRepl.Tests/EvaluationTests.cs
@@ -12,6 +12,7 @@ using CSharpRepl.Services.Dotnet;
 using CSharpRepl.Services.Roslyn;
 using CSharpRepl.Services.Roslyn.Scripting;
 using Xunit;
+using static PrettyPrompt.Highlighting.FormattedString.TextElementsEnumerator;
 
 namespace CSharpRepl.Tests;
 
@@ -222,5 +223,22 @@ public class EvaluationTests : IAsyncLifetime
         Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.CharSpanOutput)}", r3.GetType().FullName);
         Assert.Equal(11, r3.Count);
         Assert.Equal(true, r3.SpanWasReadOnly);
+    }
+
+    /// <summary>
+    /// https://github.com/waf/CSharpRepl/issues/318
+    /// </summary>
+    [Fact]
+    public async Task Evaluate_RefStructResult()
+    {
+        var e1 = await services.EvaluateAsync(@"ref struct S; default(S)");
+        var r1 = Assert.IsType<EvaluationResult.Success>(e1).ReturnValue.Value;
+        Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.RefStructOutput)}", r1.GetType().FullName);
+        Assert.Equal($"Cannot output a value of 'S' because it's a ref-struct. It has to override ToString() to see its value.", r1.ToString());
+
+        var e2 = await services.EvaluateAsync(@"ref struct S{public override string ToString()=>""custom result"";} default(S)");
+        var r2 = Assert.IsType<EvaluationResult.Success>(e2).ReturnValue.Value;
+        Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.RefStructOutput)}", r2.GetType().FullName);
+        Assert.Equal("custom result", r2.ToString());
     }
 }

--- a/CSharpRepl.Tests/EvaluationTests.cs
+++ b/CSharpRepl.Tests/EvaluationTests.cs
@@ -12,7 +12,6 @@ using CSharpRepl.Services.Dotnet;
 using CSharpRepl.Services.Roslyn;
 using CSharpRepl.Services.Roslyn.Scripting;
 using Xunit;
-using static PrettyPrompt.Highlighting.FormattedString.TextElementsEnumerator;
 
 namespace CSharpRepl.Tests;
 
@@ -210,19 +209,45 @@ public class EvaluationTests : IAsyncLifetime
         dynamic r1 = Assert.IsType<EvaluationResult.Success>(eval1).ReturnValue.Value;
         Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.SpanOutput)}", r1.GetType().FullName);
         Assert.Equal(3, r1.Count);
-        Assert.Equal(false, r1.SpanWasReadOnly);
+        Assert.Equal(typeof(Span<int>), r1.OriginalType);
 
         var eval2 = await services.EvaluateAsync(@"(ReadOnlySpan<int>)[1,2,3]");
         dynamic r2 = Assert.IsType<EvaluationResult.Success>(eval2).ReturnValue.Value;
         Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.SpanOutput)}", r2.GetType().FullName);
         Assert.Equal(3, r2.Count);
-        Assert.Equal(true, r2.SpanWasReadOnly);
+        Assert.Equal(typeof(ReadOnlySpan<int>), r2.OriginalType);
 
         var eval3 = await services.EvaluateAsync(@"""Hello World"".AsSpan()");
         dynamic r3 = Assert.IsType<EvaluationResult.Success>(eval3).ReturnValue.Value;
         Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.CharSpanOutput)}", r3.GetType().FullName);
         Assert.Equal(11, r3.Count);
-        Assert.Equal(true, r3.SpanWasReadOnly);
+        Assert.Equal(typeof(ReadOnlySpan<char>), r3.OriginalType);
+    }
+
+    /// <summary>
+    /// https://github.com/waf/CSharpRepl/issues/317
+    /// </summary>
+    [Fact]
+    public async Task Evaluate_MemoryResult()
+    {
+
+        var eval1 = await services.EvaluateAsync(@"new[]{1,2,3}.AsMemory()");
+        dynamic r1 = Assert.IsType<EvaluationResult.Success>(eval1).ReturnValue.Value;
+        Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.SpanOutput)}", r1.GetType().FullName);
+        Assert.Equal(3, r1.Count);
+        Assert.Equal(typeof(Memory<int>), r1.OriginalType);
+
+        var eval2 = await services.EvaluateAsync(@"(ReadOnlyMemory<int>)new[] { 1, 2, 3 }.AsMemory()");
+        dynamic r2 = Assert.IsType<EvaluationResult.Success>(eval2).ReturnValue.Value;
+        Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.SpanOutput)}", r2.GetType().FullName);
+        Assert.Equal(3, r2.Count);
+        Assert.Equal(typeof(ReadOnlyMemory<int>), r2.OriginalType);
+
+        var eval3 = await services.EvaluateAsync(@"""Hello World"".AsMemory()");
+        dynamic r3 = Assert.IsType<EvaluationResult.Success>(eval3).ReturnValue.Value;
+        Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.CharSpanOutput)}", r3.GetType().FullName);
+        Assert.Equal(11, r3.Count);
+        Assert.Equal(typeof(ReadOnlyMemory<char>), r3.OriginalType);
     }
 
     /// <summary>

--- a/CSharpRepl.Tests/EvaluationTests.cs
+++ b/CSharpRepl.Tests/EvaluationTests.cs
@@ -197,4 +197,30 @@ public class EvaluationTests : IAsyncLifetime
         var isWin = Environment.OSVersion.Platform == PlatformID.Win32NT;
         Assert.Equal(isWin, winRuntimeSelected);
     }
+
+    /// <summary>
+    /// https://github.com/waf/CSharpRepl/issues/318
+    /// </summary>
+    [Fact]
+    public async Task Evaluate_SpanResult()
+    {
+
+        var eval1 = await services.EvaluateAsync(@"new[]{1,2,3}.AsSpan()");
+        dynamic r1 = Assert.IsType<EvaluationResult.Success>(eval1).ReturnValue.Value;
+        Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.SpanOutput)}", r1.GetType().FullName);
+        Assert.Equal(3, r1.Count);
+        Assert.Equal(false, r1.SpanWasReadOnly);
+
+        var eval2 = await services.EvaluateAsync(@"(ReadOnlySpan<int>)[1,2,3]");
+        dynamic r2 = Assert.IsType<EvaluationResult.Success>(eval2).ReturnValue.Value;
+        Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.SpanOutput)}", r2.GetType().FullName);
+        Assert.Equal(3, r2.Count);
+        Assert.Equal(true, r2.SpanWasReadOnly);
+
+        var eval3 = await services.EvaluateAsync(@"""Hello World"".AsSpan()");
+        dynamic r3 = Assert.IsType<EvaluationResult.Success>(eval3).ReturnValue.Value;
+        Assert.EndsWith($"{nameof(__CSharpRepl_RuntimeHelper)}+{nameof(__CSharpRepl_RuntimeHelper.CharSpanOutput)}", r3.GetType().FullName);
+        Assert.Equal(11, r3.Count);
+        Assert.Equal(true, r3.SpanWasReadOnly);
+    }
 }

--- a/CSharpRepl.Tests/ObjectFormatting/CustomObjectFormattersTests.cs
+++ b/CSharpRepl.Tests/ObjectFormatting/CustomObjectFormattersTests.cs
@@ -149,6 +149,33 @@ public class CustomObjectFormattersTests : IClassFixture<RoslynServicesFixture>
     }
     #endregion
 
+    #region SpanFormatter
+    [Fact]
+    public void TestSpanFormatting()
+    {
+        Assert.Equal(
+            "Span<int>(3) { 1, 2, 3 }",
+            formatter.Format(
+                IEnumerableFormatter.Instance,
+                __CSharpRepl_RuntimeHelper.HandleSpanOutput(new[] { 1, 2, 3 }.AsSpan()),
+                Level.FirstSimple));
+
+        Assert.Equal(
+            "ReadOnlySpan<int>(3) { 1, 2, 3 }",
+            formatter.Format(
+                IEnumerableFormatter.Instance,
+                __CSharpRepl_RuntimeHelper.HandleSpanOutput([1, 2, 3]),
+                Level.FirstSimple));
+
+        Assert.Equal(
+            $"ReadOnlySpan<char>(5){Environment.NewLine}\"hello\"",
+            formatter.Format(
+                IEnumerableFormatter.Instance,
+                __CSharpRepl_RuntimeHelper.HandleSpanOutput("hello"),
+                Level.FirstSimple));
+    }
+    #endregion
+
     #region TupleFormatter
     [Theory]
     [MemberData(nameof(TupleData))]

--- a/CSharpRepl.Tests/ObjectFormatting/TestFormatter.cs
+++ b/CSharpRepl.Tests/ObjectFormatting/TestFormatter.cs
@@ -1,6 +1,5 @@
 ﻿#nullable enable
 
-using System;
 using CSharpRepl.Services;
 using CSharpRepl.Services.Roslyn.Formatting;
 using CSharpRepl.Services.Roslyn.Formatting.CustomObjectFormatters;

--- a/CSharpRepl.Tests/RoslynServicesFixture.cs
+++ b/CSharpRepl.Tests/RoslynServicesFixture.cs
@@ -1,5 +1,4 @@
-﻿using System;
-using System.Text;
+﻿using System.Text;
 using System.Threading.Tasks;
 using CSharpRepl.Services;
 using CSharpRepl.Services.Roslyn;

--- a/CSharpRepl.Tests/RoslynServicesTests.Overloads.cs
+++ b/CSharpRepl.Tests/RoslynServicesTests.Overloads.cs
@@ -2,7 +2,6 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-using System;
 using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;

--- a/CSharpRepl.Tests/ScriptArgumentTests.cs
+++ b/CSharpRepl.Tests/ScriptArgumentTests.cs
@@ -1,5 +1,4 @@
-﻿using System;
-using System.Threading.Tasks;
+﻿using System.Threading.Tasks;
 using CSharpRepl.Services;
 using CSharpRepl.Services.Roslyn;
 using CSharpRepl.Services.Roslyn.Scripting;

--- a/CSharpRepl.Tests/SymbolExplorerTests.cs
+++ b/CSharpRepl.Tests/SymbolExplorerTests.cs
@@ -2,7 +2,6 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-using System;
 using System.Threading.Tasks;
 using CSharpRepl.Services;
 using CSharpRepl.Services.Roslyn;

--- a/CSharpRepl.Tests/SyntaxHighlightingTests.cs
+++ b/CSharpRepl.Tests/SyntaxHighlightingTests.cs
@@ -2,7 +2,6 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-using System;
 using System.Linq;
 using System.Threading.Tasks;
 using CSharpRepl.Services;

